src/states/joining_node.rs

```diff
@@ -437,6 +377,19 @@ impl Bootstrapped for JoiningNode {
     }
 }
 
+impl Unapproved for JoiningNode {
+    const SEND_ACK: bool = true;
+
+    fn get_proxy_public_id(&self, proxy_name: &XorName) -> Result<&PublicId> {
```

Comment:
Can you move this and the duplicate implementation into pub trait Unapproved, and only expose a function like:
(not sure about the name).

```
fn get_unchecked_proxy_public_id(&self) -> &PublicId {
    &self.proxy_pub_id
}
```

Reply:
This is also used in `ProvingNode` where it wouldn't work, because there is no `self.proxy_pub_id`. I can try to extract the duplicate bits into a free function though.

Reply:
I tend to think it is pretty much always worth it, for maintenance to avoid duplicate code.
But I'll let you decide.